ce/spread destructure (#21)
Ada-Developers-Academy/ada-react-curriculum: 02-react-week-1/spreading-and-destructuring.md

```javascript
const greet = function greet(greeting, ...people) {
  console.log(`${greeting} ${people}`);
}

greet('Hello', 'Ada', 'Grace', 'Katherine'); // Hello Ada,Grace,Katherine
```

In the above example `greeting` takes the value `'hello'` and all the subsequent arguments are stored in the `people` array.  So rest parameters take the _rest_ of the arguments and put them into an array.  

A few things to highlight:

- A rest parameter must be the last item in a functional expression or declaration.
- A function may only have **one** rest parameter.

## The Spread Operator

So rest parameters let us take otherwise normal arguments and put them into an array.  Sometimes however you want to do the reverse.  It can be helpful to peel off elements of an array into arguments to a method.  

```javascript
const calculateBill = (appetizer, mainCourse, desert) => {
  return 1.08 * (appetizer + mainCourse + desert);
}

const prices = [3.75, 15.80, 10.0];

console.log (calculateBill(...prices)); // 31.914
```

> In the above example:  `calculateBill(...prices)` is the equivalent to doing:  `calculateBill(prices[0], prices[1], prices[2]), a much more compact syntax.

Reviewer: ```suggestion
In the above example:  `calculateBill(...prices)` is the equivalent to doing:  `calculateBill(prices[0], prices[1], prices[2])`, a much more compact syntax.
```

```

Notice that it merges the objects in order, so that `ada`'s name is overwritten by that of `language`.  It also creates an entirely **new** array which shallowly copies elements of the original object.  This means that sub-objects or arrays can be references to the original.

```javascript
const student = {
  name: 'Bob',
  grades: [3, 4, 5],
  address: {
    street: '123 Elm St',
    city: 'Trenton',
    state: 'NJ',
    zip: 11111,
   }
};

const newStudent = { ...student };

newStudent.name = 'Alice';
newStudent.address.street = '4th avenue';

console.log(newStudent);
console.log('---');
console.log(student);
```

The above code outputs: 

```javascript
{
  name: 'Alice',
  grades: [ 12, 4, 5 ],
  address: {
    street: '4th avenue',
    city: 'Trenton',
    state: 'NJ',
    zip: 11111
  }
}
---
{
  name: 'Bob',
  grades: [ 12, 4, 5 ],
  address:
   { 
     street: '4th avenue',
     city: 'Trenton',
     state: 'NJ',
     zip: 11111
  } 
}
```
<br/><br/>

**Question** Look at the code above.  Why do both Alice & Bob have a street address of 4th avenue?

<details>
  <summary>
    Click here to see the answer.
  </summary>

>   The `...` operator is making a _shallow_ copy of the original object.  It does not copy all the values of `student`'s properties.  It instead copies the _reference_ to the `address` object and the `grades` array.  To make a _deep copy_ of the original object you would need to do things [differently.](https://medium.com/@tkssharma/objects-in-javascript-object-assign-deep-copy-64106c9aefab)

Reviewer: I checked this link, the author has deleted it.  This is an alternative link.

```suggestion
  The `...` operator is making a _shallow_ copy of the original object.  It does not copy all the values of `student`'s properties.  It instead copies the _reference_ to the `address` object and the `grades` array.  To make a _deep copy_ of the original object you would need to do things [differently.](https://medium.com/technofunnel/deep-and-shallow-copy-in-javascript-110f395330c5)
```

</details>

## Examining Code

The following code simulates merging new permissions into a user account.  With a partner examine the code and answer the questions:

```javascript

> const combinePreferences = function combinePreferences(user, ...preferences) {
>   preferences.forEach((preference) => {
>     user.preferences = { ...user.preferences, ...preference}
>   });
>   return user;
> }

Author: the rest param variable name seems confusing in this context; however, if this was an intentional variable name then ignore this
```suggestion
const combinePreferences = function combinePreferences(user, ...otherPrefs) {
  otherPrefs.forEach((preference) => {
    user.preferences = { ...user.preferences, ...preference}
  });
  return user;
}
```

Author: you might need to expand the code to see the entire context

const user = {
  name: 'ada',
  preferences: {
    superuser: true,
    billableUser: false,
  }
};

const billing = {
  billableUser: true,
  getsReports: true,
  canDeployApps: false,
};

const project = {
  canCreateProjects: false,
  canDeployApps: false,
  canDeployToTest: true,
};

console.log(combinePreferences(user, billing, project));
```

**Questions** 

As a pair read the code and predict the result.  Then run the resulting code in the terminal.

> 1. Identify where the spread operator and rest parameters are.
> 2. Did it output what you expected?  Why?
> 3. Is the `forEach` loop necessary?  Could it be done without a loop?  Why or why not?

Reviewer: Just suggest making the numbered list auto numbered.

```suggestion
1. Identify where the spread operator and rest parameters are.
1. Did it output what you expected?  Why?
1. Is the `forEach` loop necessary?  Could it be done without a loop?  Why or why not?
```

<details>
  <summary>
    Click here to see some answers.
  </summary>

>   1. The rest parameter occurs at `combinePreferences(user, ...preferences)`, and the spread operator is used twice at `user.preferences = { ...user.preferences, ...preference}`.  
>   2. The result is the user object with it's preference attribute updated merging them in order of appearance.
>   3. The `forEach` loop is needed because using the spread operator on an array will result in merging in the array's keys (index numbers) and values into the preferences.  

Reviewer: ```suggestion
  1. The rest parameter occurs at `combinePreferences(user, ...preferences)`, and the spread operator is used twice at `user.preferences = { ...user.preferences, ...preference}`.  
  1. The result is the user object with it's preference attribute updated merging them in order of appearance.
  1. The `forEach` loop is needed because using the spread operator on an array will result in merging in the array's keys (index numbers) and values into the preferences.  
```
